src/mpiramsesio.cxx

```diff
@@ -98,6 +99,580 @@ void MPIDomainDecompositionRAMSES(Options &opt){
     }
 }
 
+/// Domain Decomposition by Tree
+void MPIDomainDecompositionWithTree_GetNodeID(Node *node, Int_t *impi_leafID, Int_t &index)
+{
+	if(node->GetLeaf()>0) return;
+
+	Node *left, *right;
+	left = ((SplitNode *) node)->GetLeft();
+	right= ((SplitNode *) node)->GetRight();
+
+	if(left->GetLeaf()>0 && right->GetLeaf()>0){
+		impi_leafID[index] = node->GetID();
+		index ++;
+	}
+	else
+	{
+		MPIDomainDecompositionWithTree_GetNodeID(left, impi_leafID, index);
+		MPIDomainDecompositionWithTree_GetNodeID(right, impi_leafID, index);
+	}
+}
+void MPIDomainDecompositionWithTree_NodeClose(Node *node, Int_t &node_id, Int_t &n_leaf)
+{
+	if(node->GetID() != node_id){
+		Node *left, *right;
+		if(node->GetLeaf()>0) return;
+		left = ((SplitNode *) node)->GetLeft();
+		right= ((SplitNode *) node)->GetRight();
+		MPIDomainDecompositionWithTree_NodeClose(left, node_id, n_leaf);
+		MPIDomainDecompositionWithTree_NodeClose(right, node_id, n_leaf);
+	}
+	else
+	{
+		node->SetLeaf(1);
+		n_leaf --;
+		return;
+	}
+}
+void MPIDomainDecompositionWithTree(Options &opt){
+
+	//-----
+	// Get Nbodies
+	//-----
+	long long nbodies = ReadHeader(opt);
+
+	vector<Particle> Part_mpi;
+	Part_mpi.resize(nbodies);
+	if(opt.iBaryonSearch>0){
+		cout<<"%123123-----"<<endl;
+		cout<<"%123123	MPIDomainDecompositionWithTree: Check the size of Part_mpi to include the number of baryons"<<endl;
+		cout<<"%123123-----"<<endl;
+		exit(9);
+	}
+
+	//-----
+	// Head Related
+	//-----
+        char buf[2000],buf1[2000],buf2[2000];
+        string stringbuf,orderingstring;
+        double dmp_mass, OmegaM, OmegaB;
+
+	//-----
+	// MPI Related
+	//-----
+        int *ireadfile,*ireadtask,*readtaskID;
+        Int_t Nlocalbuf,ibuf=0,*Nbuf, *Nbaryonbuf;
+
+        Nbuf=new Int_t[NProcs];
+        Nbaryonbuf=new Int_t[NProcs];
+        for (int j=0;j<NProcs;j++) Nbuf[j]=0;
+        for (int j=0;j<NProcs;j++) Nbaryonbuf[j]=0;
+
+	//-----
+	// Read Related
+	//-----
+        int idim,ivar,igrid;
+        int   typeval;
+        Int_t count2,bcount2;
+        fstream *Fpart, *Fpartmass, *Fpartage, *Fpartfam, *Famr, *Fhydro;
+        int dummy,byteoffset;
+        RAMSES_Header *header;
+        RAMSESFLOAT *xtempchunk, *mtempchunk, *agetempchunk, *xtempall;
+	int *famtempall, *dummy_family;
+	char *dummy_family2;
+	Double_t mscale,lscale,lvscale,rhoscale,scale_l,scale_d,scale_t;
+
+        int ndark  = 0;
+        int nstar  = 0;
+        int nghost = 0;
+	Int_t count_mpi = 0;
+        RAMSESFLOAT xtemp[3], ageval;
+        Double_t mtemp;
+
+        Int_t chunksize = opt.inputbufsize, nchunk;
+	xtempall   = new RAMSESFLOAT[nbodies*3];
+	famtempall = new int[nbodies];
```

Comment:
@JinsuRhee, can you explain how `xtempall` is a viable solution? It is quite possible that there are too many particles to fit in memory. Hence the previous approach to load stuff in chunks. Does it not make sense to only enable this type of load if and only if one can construct a local xtempall allocation and then build a tree to do the mpi decomposition?